### PR TITLE
Adding CPP sample for apps to use the WindowsAppSDK telemetry framework

### PR DESCRIPTION
**Pull request by ZenBird**

### CODE REVIEW

**Comment by ZenBird** [comment]
@MythiliM is added to the review. #Closed

**Comment by ZenBird** [comment]
> 

Remove

---
Refers to: Samples/WindowsAppSDKTelemetry/cpp/TelemetrySample.cpp:24 in b5caa74. [](commit_id = b5caa74eb458ba3a1dc7c2a42dbf37049abcb998, deletion_comment = False)

File: Samples/WindowsAppSDKTelemetry/cpp/TelemetrySample.cpp [hunk 1]
```diff
@@ -0,0 +1,23 @@
+// ConsoleApplication1.cpp : This file contains the 'main' function. Program execution begins and ends there.
+//
```
Comment (ZenBird) [inline Samples/WindowsAppSDKTelemetry/cpp/TelemetrySample.cpp:2]: Add WindowsAppSDK license header

File: Samples/WindowsAppSDKTelemetry/cpp/WindowsAppSDKTelemetrySample.vcxproj [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+      <SDLCheck>true</SDLCheck>
+      <PreprocessorDefinitions>NDEBUG;_CONSOLE;%(PreprocessorDefinitions)</PreprocessorDefinitions>
+      <ConformanceMode>true</ConformanceMode>
+    </ClCompile>
+    <Link>
+      <SubSystem>Console</SubSystem>
+      <EnableCOMDATFolding>true</EnableCOMDATFolding>
+      <OptimizeReferences>true</OptimizeReferences>
+      <GenerateDebugInformation>true</GenerateDebugInformation>
+      <AdditionalDependencies>$(MSBuildThisFileDirectory)..\packages\Microsoft.ProjectReunion.Foundation.0.8.1\lib\win10-$(_MrtCoreRuntimeIdentifier)\WindowsAppSDKTelemetryResourceLib.lib;%(AdditionalDependencies)</AdditionalDependencies>
```
Comment (MythiliM) [inline Samples/WindowsAppSDKTelemetry/cpp/WindowsAppSDKTelemetrySample.vcxproj:145]: Why is this needed for telemetry sample?

File: Samples/WindowsAppSDKTelemetry/cpp/WindowsAppSDKTelemetrySample.vcxproj [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+    <Error Condition="!Exists('..\packages\Microsoft.ProjectReunion.WinUI.0.8.1\build\native\Microsoft.ProjectReunion.WinUI.props')" Text="$([System.String]::Format('$(ErrorText)', '..\packages\Microsoft.ProjectReunion.WinUI.0.8.1\build\native\Microsoft.ProjectReunion.WinUI.props'))" />
+    <Error Condition="!Exists('..\packages\Microsoft.ProjectReunion.WinUI.0.8.1\build\native\Microsoft.ProjectReunion.WinUI.targets')" Text="$([System.String]::Format('$(ErrorText)', '..\packages\Microsoft.ProjectReunion.WinUI.0.8.1\build\native\Microsoft.ProjectReunion.WinUI.targets'))" />
+    <Error Condition="!Exists('..\packages\Microsoft.ProjectReunion.0.8.1\build\native\Microsoft.ProjectReunion.props')" Text="$([System.String]::Format('$(ErrorText)', '..\packages\Microsoft.ProjectReunion.0.8.1\build\native\Microsoft.ProjectReunion.props'))" />
```
Comment (MythiliM) [inline Samples/WindowsAppSDKTelemetry/cpp/WindowsAppSDKTelemetrySample.vcxproj:177]: I am not sure how many of these are actually needed. Can you please cleanup?

File: Samples/WindowsAppSDKTelemetry/cpp/WindowsAppSDKTelemetrySample.vcxproj (hunk 2 above)
Comment (ZenBird) [reply Samples/WindowsAppSDKTelemetry/cpp/WindowsAppSDKTelemetrySample.vcxproj:145]: The lib contains the definitions for the functions that load the resource from resource dll. I need to figure out if this lib needs to be standalone or can be merged with a different lib.

File: Samples/WindowsAppSDKTelemetry/cpp/TelemetrySample.cpp [hunk 4]
```diff
@@ -0,0 +1,11 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "TelemetrySample.h"
+
+int main()
+{
+    WindowsAppSDKTelemetrySample::BooleanTelemetryEvent(true);
```
Comment (MythiliM) [inline Samples/WindowsAppSDKTelemetry/cpp/TelemetrySample.cpp:8]: Can you please provide examples of how partC, PartB looks like for these examples?

**Review by MythiliM — COMMENTED** [review]
PR comments to be reviewed.

File: Samples/WindowsAppSDKTelemetry/cpp/WindowsAppSDKTelemetrySample.vcxproj (hunk 3 above)
Comment (ZenBird) [reply Samples/WindowsAppSDKTelemetry/cpp/WindowsAppSDKTelemetrySample.vcxproj:177]: This get added by default when we add a reference to the Reunion Nuget package.

File: Samples/WindowsAppSDKTelemetry/cpp/packages.config [hunk 5]
```diff
@@ -0,0 +1,9 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<packages>
+  <package id="Microsoft.ProjectReunion" version="0.8.1" targetFramework="native" />
+  <package id="Microsoft.ProjectReunion.DWrite" version="0.8.1" targetFramework="native" />
+  <package id="Microsoft.ProjectReunion.Foundation" version="0.8.1" targetFramework="native" />
+  <package id="Microsoft.ProjectReunion.InteractiveExperiences" version="0.8.1" targetFramework="native" />
+  <package id="Microsoft.ProjectReunion.WinUI" version="0.8.1" targetFramework="native" />
+  <package id="Microsoft.Windows.ImplementationLibrary" version="1.0.210204.1" targetFramework="native" />
```
Comment (MythiliM) [inline Samples/WindowsAppSDKTelemetry/cpp/packages.config:8]: Why is this needed when this is already part of reunion package?

File: Samples/WindowsAppSDKTelemetry/cpp/WindowsAppSDKTelemetrySample.vcxproj (hunk 3 above)
Comment (MythiliM) [reply Samples/WindowsAppSDKTelemetry/cpp/WindowsAppSDKTelemetrySample.vcxproj:177]: Can you remove and see if sample builds & works? Sample should be as concise as possible so consumers know what is needed.

File: Samples/WindowsAppSDKTelemetry/WindowsAppSDKTelemetry.sln [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+VisualStudioVersion = 16.0.31410.357
+MinimumVisualStudioVersion = 10.0.40219.1
+Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "WindowsAppSDKTelemetry", "cpp\WindowsAppSDKTelemetrySample.vcxproj", "{92688CEE-BB3A-4A75-9D37-A3355F7741E7}"
+EndProject
+Global
+	GlobalSection(SolutionConfigurationPlatforms) = preSolution
+		Debug|x64 = Debug|x64
+		Debug|x86 = Debug|x86
+		Release|x64 = Release|x64
+		Release|x86 = Release|x86
```
Comment (angelazhangmsft) [inline Samples/WindowsAppSDKTelemetry/WindowsAppSDKTelemetry.sln:13]: We should support ARM64 for all WinAppSDK samples as well

File: Samples/WindowsAppSDKTelemetry/WindowsAppSDKTelemetry.sln [hunk 7]
```diff
@@ -0,0 +1,31 @@
+﻿
```
Comment (angelazhangmsft) [inline Samples/WindowsAppSDKTelemetry/WindowsAppSDKTelemetry.sln:1]: Can we place the sln file under the cpp/ folder?

**Comment by angelazhangmsft** [comment]
Can we rename the folder from Samples/WindowsAppSDKTelemetry -> Samples/Telemetry? Also, a README for the sample would be helpful - there are other examples in the repo for how to structure the README.

File: Samples/WindowsAppSDKTelemetry/WindowsAppSDKTelemetry.sln (hunk 7 above)
Comment (angelazhangmsft) [reply Samples/WindowsAppSDKTelemetry/WindowsAppSDKTelemetry.sln:1]: And to follow the sample naming guidelines, we should name the folder cpp-win32

File: Samples/WindowsAppSDKTelemetry/cpp/WindowsAppSDKTelemetrySample.vcxproj (hunk 3 above)
Comment (rohanp-msft) [reply Samples/WindowsAppSDKTelemetry/cpp/WindowsAppSDKTelemetrySample.vcxproj:177]: The automatic importing of the .props and .targets files of the NuGet packages is required for the packages to work correctly in the project. These Error lines assert that.

**Comment by rohanp-msft** [comment]
In the README, would you please include instructions on how to view the telemetry? Or provide a link to a page that contains the instructions? Will the tool for viewing the telemetry logged by the app also show telemetry logged by the SDK?

File: Samples/WindowsAppSDKTelemetry/cpp/WindowsAppSDKTelemetrySample.vcxproj (hunk 2 above)
Comment (ZenBird) [reply Samples/WindowsAppSDKTelemetry/cpp/WindowsAppSDKTelemetrySample.vcxproj]: We will not need to explicitly add this when our changes are merged into the nightly build.

File: Samples/WindowsAppSDKTelemetry/cpp/packages.config (hunk 5 above)
Comment (ZenBird) [reply Samples/WindowsAppSDKTelemetry/cpp/packages.config]: WIL is not referenced by default in WindowsAppSDK yet. It may be present in the official build. Currently, the nightly does not contain the reference.

File: Samples/Insights/cpp-win32/Insights.sln [hunk 8]
```diff
@@ -0,0 +1,37 @@
+﻿
+Microsoft Visual Studio Solution File, Format Version 12.00
+# Visual Studio Version 16
+VisualStudioVersion = 16.0.31410.357
+MinimumVisualStudioVersion = 10.0.40219.1
+Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "Insights", "InsightsSample.vcxproj", "{92688CEE-BB3A-4A75-9D37-A3355F7741E7}"
+EndProject
+Global
+	GlobalSection(SolutionConfigurationPlatforms) = preSolution
+		Debug|ARM = Debug|ARM
```
Comment (angelazhangmsft) [inline Samples/Insights/cpp-win32/Insights.sln:10]: ARM64 instead of ARM :)

File: Samples/Insights/cpp-win32/InsightsSample.cpp [hunk 9]
```diff
@@ -0,0 +1,48 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "InsightsSample.h"
+
+/*
+* Boolean Event
+* 
+* "data": {
+        "value": 1,
```
Comment (rohanp-msft) [inline Samples/Insights/cpp-win32/InsightsSample.cpp:10]: Would be helpful to have comments here saying which part is Part B and which part is Part C. I keep forgetting which one the data value is.

File: Samples/Insights/cpp-win32/InsightsSample.h [hunk 10]
```diff
@@ -0,0 +1,30 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#pragma once
+
+#include "WindowsAppSDKInsights.h"
```
Comment (huichen123) [inline Samples/Insights/cpp-win32/InsightsSample.h:6]: what does this provide? All you are using is from WIL.

File: Samples/Insights/cpp-win32/InsightsSample.cpp [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+            "SDKExperimentationLevel": "",
+            "IsDebugging": false,
+            "AppSessionGuid": "00000D80-0001-0002-B425-C8C4E284D701"
+        }
+    }
+*/
+
+int main()
+{
+    InsightsSample::BooleanTelemetryEvent(true);
```
Comment (huichen123) [inline Samples/Insights/cpp-win32/InsightsSample.cpp:45]: >InsightsSample::BooleanTelemetryEvent(true); [](http://example.com/codeflow?start=4&length=44)

there is no need to register the provider?

File: Samples/Insights/cpp-win32/InsightsSample.vcxproj.filters [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+      <Extensions>cpp;c;cc;cxx;c++;cppm;ixx;def;odl;idl;hpj;bat;asm;asmx</Extensions>
+    </Filter>
+    <Filter Include="Header Files">
+      <UniqueIdentifier>{93995380-89BD-4b04-88EB-625FBE52EBFB}</UniqueIdentifier>
+      <Extensions>h;hh;hpp;hxx;h++;hm;inl;inc;ipp;xsd</Extensions>
+    </Filter>
+    <Filter Include="Resource Files">
+      <UniqueIdentifier>{67DA6AB6-F800-4c08-8B7A-83BB121AAD01}</UniqueIdentifier>
+      <Extensions>rc;ico;cur;bmp;dlg;rc2;rct;bin;rgs;gif;jpg;jpeg;jpe;resx;tiff;tif;png;wav;mfcribbon-ms</Extensions>
+    </Filter>
```
Comment (huichen123) [inline Samples/Insights/cpp-win32/InsightsSample.vcxproj.filters:15]: is there resource file?

File: Samples/Insights/cpp-win32/InsightsSample.h (hunk 10 above)
Comment (ZenBird) [reply Samples/Insights/cpp-win32/InsightsSample.h:6]: It provides the PartB information that will not be logged if you only used tracelogging.h from WIL.

File: Samples/Insights/cpp-win32/InsightsSample.cpp (hunk 11 above)
Comment (ZenBird) [reply Samples/Insights/cpp-win32/InsightsSample.cpp:45]: Nope.

File: Samples/Insights/cpp-win32/packages.config [hunk 13]
```diff
@@ -0,0 +1,5 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<packages>
+  <package id="Microsoft.Windows.ImplementationLibrary" version="1.0.210204.1" targetFramework="native" />
+  <package id="Microsoft.WindowsAppSDK.Foundation.TransportPackage" version="0.8.0-main.20210729.4.nightly" targetFramework="native" />
```
Comment (huichen123) [inline Samples/Insights/cpp-win32/packages.config:4]: >Microsoft.WindowsAppSDK.Foundation.TransportPackage [](http://example.com/codeflow?start=15&length=51)

this is internal. cannot use on public sample.

File: Samples/Insights/cpp-win32/InsightsSample.vcxproj.filters (hunk 12 above)
Comment (ZenBird) [reply Samples/Insights/cpp-win32/InsightsSample.vcxproj.filters:15]: This is the auto-generated file from Visual studio. I do not understand the comment.

Comment (huichen123) [reply Samples/Insights/cpp-win32/InsightsSample.vcxproj.filters:15]: never mind. i was just curious. was wondering why this filter exists when it's not used.

File: Samples/Insights/cpp-win32/InsightsSample.cpp (hunk 11 above)
Comment (huichen123) [reply Samples/Insights/cpp-win32/InsightsSample.cpp:45]: wondering why register is needed in OS code, but not here.

File: Samples/Insights/cpp-win32/packages.config (hunk 13 above)
Comment (ZenBird) [reply Samples/Insights/cpp-win32/packages.config:4]: Right. Until there is an official build to reference, this will have to do.

File: Samples/Insights/cpp-win32/InsightsSample.cpp [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+    InsightsSample::TextPayloadEvent(L"This text is the event payload");
+    InsightsSample::EventWithUserDefinedType(L"This is the user-defined type payload");
```
Comment (rohanp-msft) [inline Samples/Insights/cpp-win32/InsightsSample.cpp:47]: >EventWithUserDefinedType(L"This is the user-defined type payload"); [](http://example.com/codeflow?start=20&length=67)

Would you use a user defined type here, please? It'll help people that'll use this sample as a reference.

**Comment by angelazhangmsft** [comment]
> /azp run
> 
> In reply to: [895676763](https://github.com/microsoft/WindowsAppSDK-Samples/pull/50#issuecomment-895676763)

@ZenBird - we don't have a pipeline set up for this repo yet, so we are relying on manual testing currently.

File: Samples/Insights/cpp-win32/InsightsSample.cpp (hunk 14 above)
Comment (ZenBird) [reply Samples/Insights/cpp-win32/InsightsSample.cpp:47]: PCWSTR is technically a user-defined type.

File: Samples/Insights/cpp-win32/README.md [hunk 15]
```diff
@@ -0,0 +1,14 @@
+# Instructions
+This sample makes use of the Insights headers and lib to define data collection events that can be used to track different events in the application.
+
+## Collecting data locally
+The data can be locally saved by using tools.
+
+1. Use [tracelog](https://docs.microsoft.com/en-us/windows-hardware/drivers/devtest/tracelog) to start tracing
```
Comment (angelazhangmsft) [inline Samples/Insights/cpp-win32/README.md:7]: remove en-us from link

File: Samples/Insights/cpp-win32/README.md [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+The data can be locally saved by using tools.
+
+1. Use [tracelog](https://docs.microsoft.com/en-us/windows-hardware/drivers/devtest/tracelog) to start tracing
+
+2. Run your code
+
+3. Stop tracing
+
+## Viewing the data
+The above steps will produce an ETL file that contains the data. The data can be viewed using [Windows Performance analyzer](https://docs.microsoft.com/en-us/windows-hardware/drivers/devtest/capture-and-view-tracelogging-data)
```
Comment (angelazhangmsft) [inline Samples/Insights/cpp-win32/README.md:14]: remove en-us here too

File: Samples/Insights/cpp-win32/README.md [hunk 17]
```diff
@@ -0,0 +1,14 @@
+# Instructions
+This sample makes use of the Insights headers and lib to define data collection events that can be used to track different events in the application.
```
Comment (angelazhangmsft) [inline Samples/Insights/cpp-win32/README.md:2]: lib -> library?

File: Samples/Insights/cpp-win32/packages.config [hunk 18]
```diff
@@ -0,0 +1,9 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<packages>
+  <package id="Microsoft.Windows.ImplementationLibrary" version="1.0.210803.1" targetFramework="native" />
+  <package id="Microsoft.WindowsAppSDK" version="1.0.0-experimental1" targetFramework="native" />
+  <package id="Microsoft.WindowsAppSDK.DWrite" version="1.0.0-experimental1" targetFramework="native" />
+  <package id="Microsoft.WindowsAppSDK.Foundation" version="1.0.0-experimental1" targetFramework="native" />
+  <package id="Microsoft.WindowsAppSDK.InteractiveExperiences" version="1.0.0-experimental1" targetFramework="native" />
+  <package id="Microsoft.WindowsAppSDK.WinUI" version="1.0.0-experimental1" targetFramework="native" />
+</packages>
```
Comment (rohanp-msft) [inline Samples/Insights/cpp-win32/packages.config:9]: Would be good to use just the packages we need here.

File: Samples/Insights/cpp-win32/README.md [hunk 19]
```diff
@@ -0,0 +1,14 @@
...
+The data can be locally saved by using tools.
+
+1. Use [tracelog](https://docs.microsoft.com/windows-hardware/drivers/devtest/tracelog) to start tracing
+
+2. Run your code
+
+3. Stop tracing
+
+## Viewing the data
+The above steps will produce an ETL file that contains the data. The data can be viewed using [Windows Performance analyzer](https://docs.microsoft.com/windows-hardware/drivers/devtest/capture-and-view-tracelogging-data)
```
Comment (rohanp-msft) [inline Samples/Insights/cpp-win32/README.md:14]: >https://docs.microsoft.com/windows-hardware/drivers/devtest/capture-and-view-tracelogging-data [](http://example.com/codeflow?start=125&length=94)

Direct link would be better:
[https://docs.microsoft.com/en-us/windows-hardware/drivers/devtest/capture-and-view-tracelogging-data#view-tracelogging-data-using-wpa](https://docs.microsoft.com/en-us/windows-hardware/drivers/devtest/capture-and-view-tracelogging-data#view-tracelogging-data-using-wpa)

File: Samples/Insights/cpp-win32/InsightsSample.h [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+        PDT_ProductAndServiceUsage, // Privacy data data for events. Check wil/traceloggingconfig.h for a full list
+        value/* value to be logged*/);
+
+
+    // Event that contains a string as payload
+    DEFINE_COMPLIANT_TELEMETRY_EVENT_STRING(TextPayloadEvent, // Event name
+        PDT_ProductAndServiceUsage, // Privacy data data for events. Check wil/traceloggingconfig.h for a full list
+        value/* value to be logged*/);
+
+    // Event that contains a single user define data type payload
```
Comment (rohanp-msft) [inline Samples/Insights/cpp-win32/InsightsSample.h:25]: >define [](http://example.com/codeflow?start=41&length=6)

defined

File: Samples/Insights/cpp-win32/InsightsSample.h [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+
+class InsightsSample : public wil::TraceLoggingProvider
+{
+    IMPLEMENT_TRACELOGGING_CLASS(InsightsSample, "Microsoft.WindowsAppSDK.Insights.Sample",
+        // {c782669f-f4db-53f2-9674-13b8897164cc}
+        (0xc782669f, 0xf4db, 0x53f2, 0x96, 0x74, 0x13, 0xb8, 0x89, 0x71, 0x64, 0xcc));
+
+    // Event that contains a single bool value as payload
+    DEFINE_COMPLIANT_TELEMETRY_EVENT_BOOL(BooleanTelemetryEvent, // Event name
+        PDT_ProductAndServiceUsage, // Privacy data data for events. Check wil/traceloggingconfig.h for a full list
```
Comment (rohanp-msft) [inline Samples/Insights/cpp-win32/InsightsSample.h:16]: >Privacy data data for events [](http://example.com/codeflow?start=39&length=28)

Privacy data for events

File: Samples/Insights/cpp-win32/README.md [hunk 22]
```diff
@@ -0,0 +1,14 @@
+# Instructions
+This sample makes use of the Insights headers and library to define data collection events that can be used to track different events in the application.
```
Comment (rohanp-msft) [inline Samples/Insights/cpp-win32/README.md:2]: >. [](http://example.com/codeflow?start=152&length=1)

I think it would be good to make it clear to the app devs that the Insights library can't be used by them to gather telemetry from the usage of their app in the wild. That is, this data that's collected will not be pushed up to the Microsoft servers and made available to them.

Comment (ZenBird) [reply Samples/Insights/cpp-win32/README.md:2]: I don't want to introduce any divergence from our agreement in terms of the Tracelogging API. I am inclined to leave this as such. @MythiliM : Do you have any input on what must be documented?

File: Samples/Insights/cpp-win32/packages.config (hunk 18 above)
Comment (ZenBird) [reply Samples/Insights/cpp-win32/packages.config:9]: These packages come as a side-effect of adding WindowsAppSDK reference. These are not explicitly added.

Comment (rohanp-msft) [reply Samples/Insights/cpp-win32/packages.config:11]: You're referencing the metapackage. That pulls in everything. You can reference just the Foundation package.

**Comment by ZenBird** [comment]
One can add multiple providers for tracking when collecting telemetry. The tool will capture telemetry from all the providers.

---
In reply to: [888491691](https://github.com/microsoft/WindowsAppSDK-Samples/pull/50#issuecomment-888491691) [](http://example.com/codeflow?ancestors=888491691,888491691)

**Comment by angelazhangmsft** [comment]
Other comments:

- Can only build Release not Debug
- These warnings will appear when building:
- Warning	LNK4099	PDB 'Microsoft.WindowsAppSDK.Insights.pdb' was not found with 'Microsoft.WindowsAppSDK.Insights.lib(WindowsAppSDKInsights.obj)' or at 'C:\Users\angzhang\source\repos\WindowsAppSDK-Samples\Samples\Insights\cpp-win32\Release\Microsoft.WindowsAppSDK.Insights.pdb'; linking object as if no debug info	
- Warning	APPX1901	The DefaultLanguage property is either missing from the project file or does not have a value. The fallback language is set to the Visual Studio language: en-US.	Insights	C:\Users\angzhang\source\repos\WindowsAppSDK-Samples\Samples\Insights\cpp-win32\packages\Microsoft.WindowsAppSDK.Foundation.1.0.0-experimental1\build\Microsoft.ApplicationModel.Resources.PriGen.targets	1395

File: Samples/Insights/cpp-win32/InsightsSample.cpp (hunk 11 above)
Comment (ZenBird) [reply Samples/Insights/cpp-win32/InsightsSample.cpp:45]: The macros take care of registration.

File: Samples/Insights/cpp-win32/InsightsSample.cpp [hunk 23]
```diff
@@ -0,0 +1,48 @@
...
+#include "InsightsSample.h"
+
+/*
+* Boolean Event
+* 
+* "data": {
+        "value": 1,   <-- PartC
+        "COMMON_WINDOWSAPPSDK_PARAMS": {      <-- PartB
+            "SDKVersion": "0.9",
+            "SDKExperimentationLevel": "",
```
Comment (huichen123) [inline Samples/Insights/cpp-win32/InsightsSample.cpp:13]: the project is using 1.0-preview1. consider changing this to reduce confusion.

File: Samples/Insights/cpp-win32/InsightsSample.vcxproj [hunk 24]
```diff
@@ -0,0 +1,237 @@
...
+  <ItemGroup>
+    <ClInclude Include="InsightsSample.h" />
+  </ItemGroup>
+  <ItemGroup>
+    <None Include="packages.config" />
+  </ItemGroup>
+  <Import Project="$(VCTargetsPath)\Microsoft.Cpp.targets" />
+  <ImportGroup Label="ExtensionTargets">
+    <Import Project="packages\Microsoft.Windows.ImplementationLibrary.1.0.210803.1\build\native\Microsoft.Windows.ImplementationLibrary.targets" Condition="Exists('packages\Microsoft.Windows.ImplementationLibrary.1.0.210803.1\build\native\Microsoft.Windows.ImplementationLibrary.targets')" />
+    <Import Project="packages\Microsoft.WindowsAppSDK.Foundation.1.0.0-experimental1\build\native\Microsoft.WindowsAppSDK.Foundation.targets" Condition="Exists('packages\Microsoft.WindowsAppSDK.Foundation.1.0.0-experimental1\build\native\Microsoft.WindowsAppSDK.Foundation.targets')" />
```
Comment (huichen123) [inline Samples/Insights/cpp-win32/InsightsSample.vcxproj:227]: >experimental1 [](http://example.com/codeflow?start=71&length=13)

preview1

Comment (huichen123) [inline Samples/Insights/cpp-win32/InsightsSample.vcxproj:227]: >Foundation [](http://example.com/codeflow?start=122&length=10)

we shouldn't ask people import foundation targets.

File: Samples/Insights/cpp-win32/README.md (hunk 22 above)
Comment (rohanp-msft) [inline Samples/Insights/cpp-win32/README.md:2]: >headers and library [](http://example.com/codeflow?start=38&length=19)

header

File: Samples/Insights/cpp-win32/README.md [hunk 25]
```diff
@@ -0,0 +1,14 @@
+# Instructions
+This sample makes use of the Insights headers and library to define data collection events that can be used to track different events in the application.
+
+## Collecting data locally
+The data can be locally saved by using tools.
```
Comment (rohanp-msft) [inline Samples/Insights/cpp-win32/README.md:5]: >saved [](http://example.com/codeflow?start=24&length=5)

saved and viewed

File: Samples/Insights/cpp-win32/README.md [hunk 26]
```diff
@@ -0,0 +1,14 @@
...
+
+1. Use [tracelog](https://docs.microsoft.com/windows-hardware/drivers/devtest/tracelog) to start tracing
+
+2. Run your code
+
+3. Stop tracing
+
+## Viewing the data
+The above steps will produce an ETL file that contains the data. The data can be viewed using [Windows Performance analyzer](https://docs.microsoft.com/windows-hardware/drivers/devtest/capture-and-view-tracelogging-data#view-tracelogging-data-using-wpa)
```
Comment (rohanp-msft) [inline Samples/Insights/cpp-win32/README.md:14]: >analyzer [](http://example.com/codeflow?start=115&length=8)

nit: capitalize this too
